snapcraft/parts/lifecycle.py

```diff
@@ -178,10 +199,19 @@ def _run_command(
 
     lifecycle.run(step_name)
 
-    # Generate snap.yaml
+    # Extract metadata and generate snap.yaml
     project_vars = lifecycle.project_vars
     if step_name == "prime" and not part_names:
-        _update_project_metadata(project, project_vars)
+        metadata_list = lifecycle.extract_metadata()
+        update_project_metadata(
+            project,
+            project_vars=project_vars,
+            metadata_list=metadata_list,
+            assets_dir=assets_dir,
+            prime_dir=lifecycle.prime_dir,
+        )
+
+        # TODO: copy meta/gui assets
```

Comment:
I assume this is copy snap/gui to meta/gui

Reply:
It needs to do that and update the desktop file to replace the icon entry to the final icon location. This (somewhat complex) logic is already implemented in legacy (`setup_assets`, `find_icon_file`, `finalize_icon`, `write_snap_directory`) and can be ported with minor changes. This code already handles download/installation of remote icons.